api/src/controllers/users.js

```diff
@@ -222,6 +222,7 @@ module.exports = {
         warn_docs: warn,
         warn: warn >= usersService.DOC_IDS_WARN_LIMIT,
         limit: usersService.DOC_IDS_WARN_LIMIT,
+        code: res.statusCode,
```

Comment:
I think adding the statusCode as part of the body is redundant and not common practice. So I don't think we should fall back to this.

I think the fix should be on the client-side, where we assume that `utils.fetchJSON` will actually return the statusCode, when it does not, it only returns the body of the response if the response is successful: 
https://github.com/medic/cht-core/blob/1f0f990150bf2e7df0242185f87365e96dbbecf9/webapp/src/js/bootstrapper/utils.js#L35-L36

To simulate how this works, run this code replacing the url with your CouchDb local install:
```
const fetch = require('node-fetch');
const url = 'http://admin:pass@127.0.0.1:6984';

(async () => {
  const response = await fetch(url);
  console.log(response);
  const data = await response.json();
  console.log(data);

  const willError = await fetch(`${url}/thisdbdoesnotexist`);
  console.log(willError);
  const errorData = await willError.json();
  console.log(errorData);
})();
```

Reply:
Should `fetchJSON` add the status code to the data?

```json
{
   ...data,
  code: res.statusCode
}
```

Reply:
I think that the most intuitive way of doing it is to have `fetchJson` throw an error when the response is not ok. 
We can leverage [`response.ok`](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok), and throw an error that includes the error code (optionally also the response body) if the response is not ok. 
There's an example included here: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#checking_that_the_fetch_was_successful 
Here's another example of how to handle fetch errors from the PouchDb source code: https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-http/src/index.js#L217

Reply:
Does this work? The response from `fetchJSON` will return an interface similar to this:

```ts
interface FetchJSONResponse {
  code: number;
  ok: Boolean;
  data: any;
}
```

The new `fetchJSON` implementation will be similar to this:
```js
const fetchJSON = (path) => {
  const baseUrl = getBaseUrl();
  const options = {
    credentials: 'same-origin',
    headers: opts.remote_headers
  };

  const result = {};

  return fetch(`${baseUrl}${path}`, options)
    .then(res => {
      result.code = res.status;
      result.ok = res.ok;
      return res.json();
    })
    .then(data => {
      result.data = data;
      return result;
    });
};
```

The success check in the bootstrapper will check if the `response` is `ok` alongside the `code`.

Reply:
I think this will work well. 

We don't yet have a standard of handing fetch requests. We use `request` (via `request-promise-native`) server-side to make http requests, and that uses the pattern of throwing an error when the status is not in the success range. 
Because `request` is no longer supported, and frontend already uses `fetch`, we will probably migrate to `node-fetch` for server-side gradually soon. 
All of our code relies on the http request to throw an error when an non-successful status code is received, so we will probably continue with this pattern even when using `node-fetch`, to reduce the volume of refactors necessary. 

This is why I suggested you should throw an error on a non-successful status in this case as well, because we will most likely use this pattern on fetch responses server-side, and it's generally best to be consistent.